src/libsyntax/feature_gate.rs

```diff
@@ -682,36 +817,37 @@ pub fn check_for_pushpop_syntax(f: Option<&Features>, diag: &Handler, span: Span
 }
 
 struct Context<'a> {
-    features: Vec<&'static str>,
+    features: Features,
     span_handler: &'a Handler,
     cm: &'a CodeMap,
     plugin_attributes: &'a [(String, AttributeType)],
 }
 
-impl<'a> Context<'a> {
-    fn enable_feature(&mut self, feature: &'static str) {
-        debug!("enabling feature: {}", feature);
-        self.features.push(feature);
-    }
-
-    fn gate_feature(&self, feature: &str, span: Span, explain: &str) {
-        let has_feature = self.has_feature(feature);
-        debug!("gate_feature(feature = {:?}, span = {:?}); has? {}", feature, span, has_feature);
-        if !has_feature && !self.cm.span_allows_unstable(span) {
-            emit_feature_err(self.span_handler, feature, span, GateIssue::Language, explain);
+macro_rules! gate_feature_fn {
+    ($cx: expr, $has_feature: expr, $span: expr, $name: expr, $explain: expr) => {{
+        let (cx, has_feature, span, name, explain) = ($cx, $has_feature, $span, $name, $explain);
```

Comment:
Previously we had a `gate_feature` feature function that took a feature name as a string, an span, and an error message and would check that the feature was enabled, and, if not, report an error message at that span. Now that feature names are attributes, it must be a macro. This specific line is to avoid duplication of expressions. It might not be necessary in this specific case since all the invocations of this macro are pretty simple, but I have the habit of doing that when I write macros. In more complex cases, not doing so might result in duplicating computation (which can be very bad in the presence of side effects) or in the case of Rust, borrow-checker errors, I guess.